app_dart/lib/src/request_handlers/get_timeseries_history.dart

```diff
@@ -0,0 +1,117 @@
+
+import 'package:appengine/appengine.dart';
+import 'package:cocoon_service/src/model/appengine/benchmark_data.dart';
+import 'package:cocoon_service/src/model/appengine/time_series_value.dart';
+import 'package:gcloud/db.dart';
+import 'package:meta/meta.dart';
+
+import '../datastore/cocoon_config.dart';
+import '../model/appengine/commit.dart';
+import '../model/appengine/key_helper.dart';
+import '../model/appengine/time_series.dart';
+import '../model/appengine/time_series_entity.dart';
+import '../request_handling/body.dart';
+import '../request_handling/exceptions.dart';
+import '../request_handling/regular_request_handler.dart';
+import '../service/datastore.dart';
+
+@immutable
+class GetTimeSeriesHistory extends RegularRequestHandler<GetTimeSeriesHistoryResponse> {
+  const GetTimeSeriesHistory(
+    Config config, {
+    @visibleForTesting DatastoreServiceProvider datastoreProvider,
+  })  : datastoreProvider =
+            datastoreProvider ?? DatastoreService.defaultProvider,
+        super(config: config);
+
+  final DatastoreServiceProvider datastoreProvider;
+  static const String timeSeriesKeyParam = 'TimeSeriesKey';
+  static const String startFromParam = 'StartFrom';
+
+  @override
+  Future<GetTimeSeriesHistoryResponse> post() async {
+    checkRequiredParameters(<String>[timeSeriesKeyParam]);
+    const int maxRecords = 1500;
+    final DatastoreService datastore = datastoreProvider();
+    final KeyHelper keyHelper = KeyHelper(applicationContext: AppEngineContext(false,'','','','','',Uri()));
+    final Set<Commit> commits =
+        await datastore.queryRecentCommits(limit: maxRecords).toSet();
+
+    Key timeSeriesKey;
+    try {
+      timeSeriesKey = keyHelper.decode(requestData[timeSeriesKeyParam]);
+    } catch (error) {
+      throw BadRequestException(
+          'Bad timeSeries key: ${requestData[timeSeriesKeyParam]}');
+    }
+
+    final TimeSeries timeSeries =
+        await config.db.lookupValue<TimeSeries>(timeSeriesKey, orElse: () {
+      throw BadRequestException('No such timeseries: ${timeSeriesKey.id}');
+    });
+
+    List<TimeSeriesValue> timeSeriesValues = await datastore
+        .queryRecentTimeSeriesValues(timeSeries,
+            startFrom: startFromParam, limit: maxRecords)
+        .toList();
+    timeSeriesValues = insertMissingTimeseriesValues(timeSeriesValues, commits);
+
+    return GetTimeSeriesHistoryResponse(timeSeries, timeSeriesValues);
+  }
+}
+
+@immutable
+class GetTimeSeriesHistoryResponse extends JsonBody {
+  const GetTimeSeriesHistoryResponse(this.timeSeries, this.timeSeriesValues)
+      : assert(timeSeries != null);
+
+  final TimeSeries timeSeries;
+  final List<TimeSeriesValue> timeSeriesValues;
+
+  @override
+  Map<String, dynamic> toJson() {
+    return <String, dynamic>{
+      'BenchmarkData': BenchmarkData(
+          timeSeriesEntity: TimeseriesEntity(timeSeries: timeSeries),
+          values: timeSeriesValues),
+      //LastPosition to do @Keyong
```

Comment:
What does this comment mean?

Reply:
Updated to
// TODO(keyonghan): implemement last position
// flutter/flutter#42362